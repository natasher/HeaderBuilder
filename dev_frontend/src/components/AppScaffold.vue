<template>
  <wrapper-clearfix
    id             = "mfn-wrapper"
    class          = "mfn-hb"
    :data-device   = "getCurrentView"
    :data-position = "getLayoutPosition || 'top'"
    :data-type     = "getStickyFlag"
    v-cloak >

    <wrapper-clearfix id = "mfn-hb-header">

      <!-- @slot Place App header in here -->
      <slot name = "header" />

    </wrapper-clearfix>

    <!-- @slot App main part, grid etc. -->
    <slot name = "main" />

    <modals-list />
  </wrapper-clearfix>
</template>

<script>
import { mapState, mapGetters } from 'vuex'
import ModalsList      from './ModalsList.vue'
import WrapperClearfix from './BaseGeneric/WrapperClearfix.vue'

export default {

  name: 'AppScaffold',

  components: {
    ModalsList,
    WrapperClearfix,
  },

  computed: {
    ...mapState( 'general', {
      getCurrentView: state => state.currentView,
    }),
    ...mapState( 'ui', {
      getStickyFlag: state => state.stickyFlag,
    }),
    ...mapGetters( 'ui', [ 'getLayoutPosition' ]),
  },
}
</script>
