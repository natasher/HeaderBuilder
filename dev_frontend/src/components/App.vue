<template>
  <app-scaffold>

    <base-header slot = "header" />

    <section slot = "main" id = "mfn-hb-main">

      <grid-toolbar />

      <grid />

      <inactive-items v-if = "getCurrentView === 'mobile'" />
      <items-tray    v-once />
      <footer-links  v-once />

    </section>

  </app-scaffold>
</template>

<script>
import { mapState, mapActions } from 'vuex'
import AppScaffold   from './AppScaffold.vue'
import Grid          from './Grid/Grid.vue'
import BaseHeader    from './BaseHeader/BaseHeader.vue'
import GridToolbar   from './BaseHeader/GridToolbar.vue'
import InactiveItems from './Grid/InactiveItems.vue'
import ItemsTray     from './ItemsTray.vue'
import FooterLinks   from './FooterLinks.vue'

export default {

  name: 'App',

  components: {
    AppScaffold,
    Grid,
    BaseHeader,
    GridToolbar,
    InactiveItems,
    ItemsTray,
    FooterLinks,
  },

  computed: mapState( 'general', {
    getCurrentView: state => state.currentView,
  }),

  methods: {
    ...mapActions( 'endpoint', [
      'pushWPMenusList',
      'getStateFromWP',
    ]),

    ...mapActions( 'ui', [
      'changeDevice',
    ]),

    ...mapActions( 'modals', [ 'showLaunchModal' ])
  },

  /** Lifecycle methods: */
  beforeMount() {
    /**
     * AJAX load data to state
     */
    this.pushWPMenusList()
    this.getStateFromWP()
    this.changeDevice({ device: 'desktop' })
  },

  mounted() {
    this.showLaunchModal({ $modal: this.$modal })
  },

}
</script>