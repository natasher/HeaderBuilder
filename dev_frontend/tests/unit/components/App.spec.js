import { mount, createLocalVue } from '@vue/test-utils'
import Vuex          from 'vuex'
import flushPromises from 'flush-promises'
import merge         from 'lodash.merge'
import App           from '../../../src/components/App.vue'
import AppScaffold   from '../../../src/components/AppScaffold'

const localVue = createLocalVue()
localVue.use( Vuex )

function createStore(overrides) {
  const defaultStoreConfig = {
    modules: {
      general: {
        namespaced: true,
        state: {
          currentView: 'desktop'
        }
      },
      endpoint: {
        namespaced: true,
        actions: {
          pushWPMenusList: jest.fn(() => Promise.resolve()),
          getStateFromWP : jest.fn(() => Promise.resolve()),
        }
      },
      ui: {
        namespaced: true,
        state: {
          stickyFlag: 'default'
        },
        actions: {
          changeDevice: jest.fn(() => Promise.resolve()),
        },
        getters: {
          getLayoutPosition: jest.fn(() => { return 'top' })
        }
      },
      modals: {
        namespaced: true,
        actions: {
          showLaunchModal: jest.fn(),
        }
      }
    }
  }

  return new Vuex.Store(
    merge( defaultStoreConfig, overrides )
  )

}

function createWrapper(overrides) {
  const defaultMountingOptions = {
    localVue,
    stubs: [
      'Grid',
      'BaseHeader',
      'GridToolbar',
      'InactiveItems',
      'ItemsTray',
      'FooterLinks',
      'ModalsList',
      'ResetBuilder',
    ],
    mocks: {
      $modal: jest.fn()
    },
    store: createStore()
  }

  return mount( App, merge( defaultMountingOptions, overrides ))
}

describe('App.vue', () => {

  describe('snapshots', () => {
    test('should dispay `inactive items` if current device is `mobile`', async () => {
      const currentView = 'mobile'
      const store = createStore({
        modules: {
          general: {
            state: {
              currentView
            }
          },
        }
      })
      const wrapper = createWrapper({ store, })
      await flushPromises()

      expect( wrapper.findAll( AppScaffold ).length ).toBe( 1 )
      expect( wrapper.element ).toMatchSnapshot()
    })

    test('should match snapshot if current device is NOT `mobile`', async () => {
      const wrapper = createWrapper()
      await flushPromises()

      expect( wrapper.findAll( AppScaffold ).length ).toBe( 1 )
      expect( wrapper.element ).toMatchSnapshot()
    })
  })

  test('should push fetched Wordpress menus to the state', async () => {
    const pushWPMenusList = jest.fn(() => Promise.resolve())
    const store = createStore({
      modules: {
        endpoint: {
          namespaced: true,
          actions: {
            pushWPMenusList
          }
        }
      }
    })
    createWrapper({ store, })
    await flushPromises()

    expect( pushWPMenusList ).toBeCalled()
  })

  test('should set fetched Be & WP fonts list to the state', async () => {
    const setFontsList = jest.fn(() => Promise.resolve())
    const store = createStore({
      modules: {
        endpoint: {
          namespaced: true,
          actions: {
            setFontsList
          }
        }
      }
    })
    createWrapper({ store, })
    await flushPromises()

    expect( setFontsList ).toBeCalled()
  })

  test('should settle fetched Header Builder state by ajax', async () => {
    const getStateFromWP = jest.fn(() => Promise.resolve())
    const store = createStore({
      modules: {
        endpoint: {
          namespaced: true,
          actions: {
            getStateFromWP
          }
        }
      }
    })
    createWrapper({ store, })
    await flushPromises()

    expect( getStateFromWP ).toBeCalled()
  })

  test('should set `desktop` device on load', async () => {
    const changeDevice = jest.fn(() => Promise.resolve())
    const store = createStore({
      modules: {
        ui: {
          actions: {
            changeDevice,
          }
        }
      }
    })
    createWrapper({ store, })
    await flushPromises()

    expect( changeDevice ).toBeCalledWith( expect.anything(), { device: 'desktop' }, void 0 )
  })

  test('should open beta info modal when App.vue has mounted', async () => {
    const showLaunchModal = jest.fn()
    const store = createStore({
      modules: {
        modals: {
          actions: {
            showLaunchModal,
          }
        }
      }
    })
    const mocks = {
      $modal: jest.fn()
    }

    createWrapper({ store, mocks, })
    await flushPromises()

    expect(showLaunchModal).toBeCalledWith(
      expect.anything(),
      { $modal: mocks.$modal },
      void 0
    )
  })

})