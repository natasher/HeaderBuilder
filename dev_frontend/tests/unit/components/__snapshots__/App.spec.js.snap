// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App.vue snapshots should dispay \`inactive items\` if current device is \`mobile\` 1`] = `
<div
  class="clearfix mfn-hb"
  data-device="mobile"
  data-position="top"
  data-type="default"
  id="mfn-wrapper"
>
  <div
    class="clearfix"
    id="mfn-hb-header"
  >
    <baseheader-stub />
  </div>
   
  <section
    id="mfn-hb-main"
  >
    <gridtoolbar-stub />
     
    <grid-stub />
     
    <inactiveitems-stub />
     
    <itemstray-stub />
     
    <footerlinks-stub />
  </section>
   
  <modalslist-stub />
</div>
`;

exports[`App.vue snapshots should match snapshot if current device is NOT \`mobile\` 1`] = `
<div
  class="clearfix mfn-hb"
  data-device="desktop"
  data-position="top"
  data-type="default"
  id="mfn-wrapper"
>
  <div
    class="clearfix"
    id="mfn-hb-header"
  >
    <baseheader-stub />
  </div>
   
  <section
    id="mfn-hb-main"
  >
    <gridtoolbar-stub />
     
    <grid-stub />
     
    <!---->
     
    <itemstray-stub />
     
    <footerlinks-stub />
  </section>
   
  <modalslist-stub />
</div>
`;
